WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift

```diff
@@ -105,15 +105,65 @@ private extension DashboardViewController {
         navigationItem.backBarButtonItem = backButton
     }
 
-    func configureDashboardUI() {
+    func reloadDashboardUIStatsVersion() {
         // A container view is added to respond to safe area insets from the view controller.
         // This is needed when the child view controller's view has to use a frame-based layout
         // (e.g. when the child view controller is a `ButtonBarPagerTabStripViewController` subclass).
         view.addSubview(containerView)
         containerView.translatesAutoresizingMaskIntoConstraints = false
         view.pinSubviewToSafeArea(containerView)
 
-        dashboardUI = DashboardUIFactory.dashboardUI()
+        guard let siteID = ServiceLocator.stores.sessionManager.defaultStoreID else {
+            return
+        }
+
+        DashboardUIFactory.dashboardUIStatsVersion(isFeatureFlagOn: FeatureFlag.stats.enabled,
+                                                   siteID: siteID,
+                                                   onInitialUI: { [weak self] statsVersion in
+                                                    self?.updateDashboardUI(statsVersion: statsVersion)
+            },
+                                                   onUpdate: { [weak self] statsVersion in
+                                                    self?.onDashboardUIUpdate(statsVersion: statsVersion)
+        })
+    }
+}
+
+// MARK: - Updates
+//
+private extension DashboardViewController {
+    func onDashboardUIUpdate(statsVersion: StatsVersion) {
```

Comment:
So if I understand correctly, we pass a completion block to DashboardUIFactory, that would executioner's this method here, which would check then some other logic to decide what UI to present... why not do all of that in DashboardUIFactory?

That way we'd remove the logic that decides what to present from the UI, which would also help with testability

Reply:
It looks as well like we'd be tracking the last version in both the view controller and the DashboardUIFactory?

Reply:
> It looks as well like we'd be tracking the last version in both the view controller and the DashboardUIFactory?

did you mean analytics tracking or user defaults?

Reply:
> It looks as well like we'd be tracking the last version in both the view controller and the DashboardUIFactory?

I see what you meant now, where I track the last stats and the updated stats versions in both the view controller and UI factory. With the UX in mind, I think it'd be better for the view controller to decide how to handle stats version changes. I will remove the stats version tracking from the factory. Lemme know if I missed anything!

Reply:
> we don't want to just switch the dashboard to v4 UI. We'll show a banner at the top for the user to opt in v4. On the other hand, when we're currently showing stats v4 and now stats v4 is unavailable, we just revert the dashboard UI to v3 and show a banner with explanations.

But that logic can still be encapsulated in the DashboardUIFactory, right? After all, DashboardUIFactory is not a factory anymore. It does not return any UI, but executes a callback.

Let me know if I'm missing something but what I understand from the code is that we'd need to use the DashboardUIFacotry anyway to check if v3 or v4 are available, and that once the check is completed, we either present one UI or the other (and that responsibility is in the vc at the moment, as the factory does not return a UI?). So, in theory it should be possible to make the factory request from the view controller to present the banner? or even better, the DashboardUIFactory could present the banner itself? (or even better delegate it to a different class).

I am not completely sure that makes sense though 😊